File: DSA/Deeptansh-nagarr/day 11/question1.java
import java.util.Scanner;

class PriorityQueue {
    private int[] heap;
    private int size;

    public PriorityQueue(int capacity) {
        heap = new int[capacity + 1];
        size = 0;
    }

    public void insert(int value) {
        if (size == heap.length - 1) {
            resize();
        }
        heap[++size] = value;
        int index = size;
        while (index > 1 && heap[index / 2] > heap[index]) {
            swap(index, index / 2);
            index = index / 2;
        }
    }

    public int extractMin() {
        if (size == 0) {
            throw new RuntimeException("Priority queue is empty");
        }
        int min = heap[1];
        heap[1] = heap[size--];
        heapify(1);
        return min;
    }

    private void resize() {
        int[] newHeap = new int[heap.length * 2];
        for (int i = 0; i < heap.length; i++) {
            newHeap[i] = heap[i];
        }
        heap = newHeap;
    }

    private void swap(int i, int j) {
        int temp = heap[i];
        heap[i] = heap[j];
        heap[j] = temp;
    }

    private void heapify(int index) {
        int smallest = index;
        int left = 2 * index;
        int right = 2 * index + 1;

        if (left <= size && heap[left] < heap[smallest]) {
            smallest = left;
        }

        if (right <= size && heap[right] < heap[smallest]) {
            smallest = right;
        }
      
        if (smallest!= index) {
            swap(index, smallest);
            heapify(smallest);
        }
    }
}

public class Main {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        System.out.println("Enter the capacity of the priority queue:");
        int capacity = scanner.nextInt();
        PriorityQueue priorityQueue = new PriorityQueue(capacity);
        while (true) {
            System.out.println("1. Insert element");
            System.out.println("2. Extract minimum element");
            System.out.println("3. Exit");
            System.out.println("Enter your choice:");
            int choice = scanner.nextInt();
            switch (choice) {
                case 1:
                    System.out.println("Enter the element to insert:");
                    int element = scanner.nextInt();
                    priorityQueue.insert(element);
                    break;
                case 2:
                    try {
                        int min = priorityQueue.extractMin();
                        System.out.println("Extracted minimum element: " + min);
                    } catch (RuntimeException e) {
                        System.out.println(e.getMessage());
                    }
                    break;
                case 3:
                    System.exit(0);
                    break;
                default:
                    System.out.println("Invalid choice");
            }
        }
    }
}

File: DSA/Devansh-Rawat/Day-11/Question-1.cpp
#include <iostream>
#include <vector>
#include <stdexcept>
using namespace std;
class PriorityQueue {
public:
    PriorityQueue() {}
    void insert(int element) {
        heap.push_back(element);
        siftUp(heap.size() - 1);
    }
    int extractMin() {
        if (heap.empty()) {
            throw out_of_range("Priority queue is empty");
        }
        int minElement = heap[0];
        heap[0] = heap.back();
        heap.pop_back();
        if (!heap.empty()) {
            siftDown(0);
        }
        return minElement;
    }
    int getMin() const {
        if (heap.empty()) {
            throw out_of_range("Priority queue is empty");
        }
        return heap[0];
    }
    bool isEmpty() const {
        return heap.empty();
    }
private:
    vector<int> heap;
    void siftUp(int i) {
        while (i > 0) {
            int parent = (i - 1) / 2;
            if (heap[i] >= heap[parent]) {
                break;
            }
            std::swap(heap[i], heap[parent]);
            i = parent;
        }
    }
    void siftDown(int i) {
        int leftChild, rightChild, smallest, n = heap.size();
        while (true) {
            leftChild = 2 * i + 1;
            rightChild = 2 * i + 2;
            smallest = i;
            if (leftChild < n && heap[leftChild] < heap[smallest]) {
                smallest = leftChild;
            }
            if (rightChild < n && heap[rightChild] < heap[smallest]) {
                smallest = rightChild;
            }
            if (smallest == i) {
                break;
            }
            swap(heap[i], heap[smallest]);
            i = smallest;
        }
    }
};
int main() {
    PriorityQueue pq;
    pq.insert(3);
    pq.insert(1);
    pq.insert(4);
    pq.insert(1);
    pq.insert(5);
    pq.insert(9);
    cout << "Extracting minimum elements:" << endl;
    while (!pq.isEmpty()) {
        cout << pq.extractMin() << " ";
    }
    cout << endl;
    return 0;
}
